titus-ext/jooq/src/main/java/com/netflix/titus/ext/jooq/relocation/JooqTaskRelocationGC.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    /**
+     * Removes all entries older than the given time threshold.
+     */
+    @VisibleForTesting
+    int removeExpiredData(long timeThreshold) {
+        // Count all items
+        int allCount = dslContext.fetchCount(RELOCATION_STATUS);
+        logger.info("All rows in 'relocation_status' table: {}", allCount);
+        allRowsGauge.set(allCount);
+
+        int expiredCount = dslContext.fetchCount(
+                RELOCATION_STATUS,
+                RELOCATION_STATUS.RELOCATION_EXECUTION_TIME.lt(Timestamp.from(Instant.ofEpochMilli(timeThreshold)))
+        );
+        logger.info("Expired rows in 'relocation_status' table: {}", expiredCount);
+        expiredRowsGauge.set(allCount);
+
+        if (expiredCount <= 0) {
+            return 0;
+        }
+
+        // Locate timestamp from which to remove.
+        Result<Record2<String, Timestamp>> timestampRow = dslContext.select(RELOCATION_STATUS.TASK_ID, RELOCATION_STATUS.RELOCATION_EXECUTION_TIME)
+                .from(RELOCATION_STATUS)
+                .where(RELOCATION_STATUS.RELOCATION_EXECUTION_TIME.lt(Timestamp.from(Instant.ofEpochMilli(timeThreshold))))
+                .orderBy(RELOCATION_STATUS.RELOCATION_EXECUTION_TIME.asc())
+                .limit(configuration.getGcRowLimit())
+                .fetch();
+
+        if (timestampRow.isEmpty()) {
+            logger.info("No expired data found");
+            return 0;
+        }
+        Timestamp lastToRemove = timestampRow.get(timestampRow.size() - 1).getValue(RELOCATION_STATUS.RELOCATION_EXECUTION_TIME);
+
+        // Load all data up to the given timestamp. This could be more data than above (and more than the GC limit when
+        // there are records with the lastToRemove timestamp, which were not returned due to the limit constraint.
+        // This is fine, as we do not expect that there are too many like this.
+        Result<Record2<String, Timestamp>> toRemoveRows = dslContext.select(RELOCATION_STATUS.TASK_ID, RELOCATION_STATUS.RELOCATION_EXECUTION_TIME)
+                .from(RELOCATION_STATUS)
+                .where(RELOCATION_STATUS.RELOCATION_EXECUTION_TIME.le(lastToRemove))
+                .orderBy(RELOCATION_STATUS.RELOCATION_EXECUTION_TIME.asc())
+                .fetch();
+
+        List<Pair<String, Long>> toRemoveSet = toRemoveRows.stream()
+                .map(r -> Pair.of(r.get(RELOCATION_STATUS.TASK_ID), r.get(RELOCATION_STATUS.RELOCATION_EXECUTION_TIME).getTime()))
+                .collect(Collectors.toList());
+        logger.info("Records to remove: {}", toRemoveSet);
+
+        int removedFromDb = dslContext.delete(RELOCATION_STATUS).where(RELOCATION_STATUS.RELOCATION_EXECUTION_TIME.le(lastToRemove)).execute();
+
+        logger.info("Removed expired rows from 'relocation_status' table: {}", removedFromDb);
+        gcCounter.increment(allCount);
+
+        relocationResultStore.removeFromCache(timeThreshold);
```

Comment:
minor (feel free to ignore): we have all the ids in `toRemoveSet` above, could invalidate only those to avoid unnecessary cache misses